### PR TITLE
fix: fouled attachments not being cleaned along with the gun

### PR DESCRIPTION
**Pull request by YeOldeMiller**
## Summary
SUMMARY: Bugfixes "Fix attachments not being cleaned along with the gun"

## Purpose of change

Gunmods such as underbarrel shotguns have their own fouling stat, independent of the gun they're mounted on, but no way to clean them.

## Describe the solution

Amend the code that resets `"dirt"` stat of the main gun, have it iterate over attached gunmods and reset it for any that also have that parameter.

## Describe alternatives you've considered

None that I could think of. 

## Testing

Loaded a game where I had a LeMat with heavily fouled shotgun barrel. Cleaned the gun, made sure the attachment no longer had the "_" fouling indicator. Kept shooting the shotgun attachment until it appeared again to make sure fouling keeps accumulating normally.

## Additional context

Ideally we should also have gunmod fouling enable the Mend action, but currently only main gun is checked. If one were to exclusively shoot their underbarrel attachment but not the main gun, they wouldn't be able to clean it. I want to keep the fix as simple as possible so not going to address such an extreme edge case at this time.

### CODE REVIEW

**Review by chaosvolt — COMMENTED** [review]
Errors look to be unrelated I think, unless this test failure is somehow:
```
([slow] ~starting_items)=> -------------------------------------------------------------------------------
([slow] ~starting_items)=> Aiming at a target partially covered by a wall
([slow] ~starting_items)=> -------------------------------------------------------------------------------
([slow] ~starting_items)=> ../tests/ranged_aiming_test.cpp:225
([slow] ~starting_items)=> ...............................................................................
([slow] ~starting_items)=> 
([slow] ~starting_items)=> ../tests/ranged_aiming_test.cpp:265: FAILED:
([slow] ~starting_items)=>   CHECK( unseen == 0 )
([slow] ~starting_items)=> with expansion:
([slow] ~starting_items)=>   2500 (0x9c4) == 0
([slow] ~starting_items)=> with messages:
([slow] ~starting_items)=>   unseen := 2500 (0x9c4)
([slow] ~starting_items)=>   failed := {  }
```

**Review by scarf005 — CHANGES_REQUESTED** [review]
image

https://github.com/cataclysmbnteam/Cataclysm-BN/assets/54838975/81fe16ba-db7d-4583-97f4-52c49f0aea78

1. spawned `M1911` and `Lemat Revolver Shotgun`.
2. attached and only fired `Lemat Revolver Shotgun` until it accumulated foulings.
3. tried to apply gunsmith repair kit.
4. was unable to mend attachment.

I think the logic to check whether it's possible to mend guns should be extended to check gunmods as well.

**Comment by YeOldeMiller** [comment]
> I think the logic to check whether it's possible to mend guns should be extended to check gunmods as well.

Ugh I really didn't want to go any deeper than absolutely necessary with this. Not feeling confident at all in my C++ skills.
I think I managed to cobble something working together but please check if the code looks okay to you.

File: src/avatar_functions.cpp [hunk 1]
```diff
@@ -208,7 +208,12 @@ void mend_item( avatar &you, item &obj, bool interactive )
     };
 
     std::vector<mending_option> mending_options;
-    for( const fault_id &f : obj.faults ) {
+
+    auto &faults( obj->faults );
+    for( const auto &mod : obj->gunmods() ) {
+        faults.insert( mod->faults.begin(), mod->faults.end() );
+    }
```
Comment (scarf005) [inline src/avatar_functions.cpp:215]: i think this would make gunmod faults to be copied into guns.

Comment (YeOldeMiller) [reply src/avatar_functions.cpp:215]: That's exactly what happened. That actually worked, in a way, but obviously not a good solution.

**Review by scarf005 — APPROVED** [review]
https://github.com/cataclysmbnteam/Cataclysm-BN/assets/54838975/ac8fc241-c06f-4941-bd3f-8f244b718e12

gun was intact, gunmod had fouls. was able to clean gunmods. however, i only tested narrow use cases, so cc @chaosvolt

**Comment by chaosvolt** [comment]
Aight, will check in a bit when I get back on desktop.

**Review by chaosvolt — COMMENTED** [review]
1. Compiled and load-tested.
2. Put together an M4 with a masterkey shotgun, both loaded with blackpowder rounds.
3. Switched to the masterkey.
4. Fired off the masterkey until it started to foul.
5. Spawned in the materials to mend the fault and did so.
6. Detached the masterkey and moved it to a new M4.
7. Activated pipe cleaner, it crashed partway through with:
image

Fun bonus annoyance: the sidebar and inventory menu all insist that the item I'm using it an M4A1 loaded with .223 remington, but hitting `f` confirms it's set to use the masterkey.

Additional unrelated problem: Putting a fouled gunmod on a clean gun and removing it also contaminates the gun with fouling.

Gonna compile in release instead of debug and see if it does it again, and if so whether I can get anything out of call stack.

**Review by chaosvolt — APPROVED** [review]
Okaaaayyy. I loaded the save and because it autosaved midway through cleaning it just...finished that with no issues.
image

Turned off autosave since it was annoying, repeated the steps of fouling the gunmod before transferring it to a clean gun, cleaned it and it cleans both the clean gun and the gunmod, no errors or crashing this time.

I'm gonna chalk this up to either "compiling in debug instead of release was a fucking mistake" or some random crash that can't be reliably reproduced.
